pkg/util/virtualmachine/virtualmachine.go

```diff
@@ -0,0 +1,53 @@
+package virtualmachine
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+
+	ctlkubevirtv1 "github.com/harvester/harvester/pkg/generated/controllers/kubevirt.io/v1"
+)
+
+// IsVMStopped checks VM is stopped or not. It will check two cases
+// 1. VM is stopped from GUI
+// 2. VM is stopped inside VM
+// These two cases are all stopped case.
+func IsVMStopped(
+	vm *kubevirtv1.VirtualMachine,
+	vmiCache ctlkubevirtv1.VirtualMachineInstanceCache,
+) (bool, error) {
+	strategy, err := vm.RunStrategy()
+	if err != nil {
+		return false, fmt.Errorf("error getting run strategy for vm %s in namespace %s: %v", vm.Name, vm.Namespace, err)
+	}
+
+	// VM is stopped from GUI
+	if strategy == kubevirtv1.RunStrategyHalted {
+		return true, nil
+	}
+
+	if vm.Status.PrintableStatus != kubevirtv1.VirtualMachineStatusStopped {
+		return false, nil
+	}
+
+	// When vm is stopped inside VM, the vmi will not be deleted.
+	// The status of vmi will be "Succeeded".
+	vmi, err := vmiCache.Get(vm.Namespace, vm.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("error getting vmi %s in namespace %s: %v", vm.Name, vm.Namespace, err)
+	}
+
+	if vmi == nil {
```

Comment:
I'd suggest we don't rely on checking vmi, but use the NotFound error as the condition. We can combine the statement to line 38.